x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/crypto/CryptoService.java

```diff
@@ -99,12 +110,83 @@ public CryptoService(Settings settings) throws IOException {
             try {
                 encryptionKey = encryptionKey(systemKey, keyLength, keyAlgorithm);
             } catch (NoSuchAlgorithmException nsae) {
-                throw new ElasticsearchException("failed to start crypto service. could not load encryption key", nsae);
+                throw new ElasticsearchException(COULD_NOT_LOAD_KEY_MESSAGE, nsae);
+            }
+        }
+
+        assertEncryptionKeyNotNull();
+    }
+
+    /**
+     * Creates a {@link CryptoService} using a secure setting as the contents of the system key.
+     *
+     * @param settings the system settings
+     * @param encryptionKeySetting the secure setting representing the system key
+     * @return a {@link CryptoService} object
+     */
+    public static CryptoService createFromEncryptionKeySetting(Settings settings, Setting<SecureString> encryptionKeySetting) {
+        if (encryptionKeySetting.exists(settings) == false) {
+            throw new ElasticsearchException(format("setting [%s] must be set in keystore", encryptionKeySetting.getKey()));
+        }
+
+        try (SecureString key = encryptionKeySetting.get(settings)) {
+            if (key == null || key.isEmpty()) {
+                throw new ElasticsearchException(
+                    format("setting [%s] contents was empty, it must be set in keystore", encryptionKeySetting.getKey())
+                );
             }
+
+            SecretKey systemKey = createSecretKey(key.getChars());
+            return new CryptoService(settings, systemKey);
+        }
+    }
+
+    private CryptoService(Settings settings, SecretKey systemKey) {
+        this.encryptionAlgorithm = ENCRYPTION_ALGO_SETTING.get(settings);
+        final int keyLength = getKeyLength(settings);
+        this.ivLength = keyLength / 8;
+        String keyAlgorithm = ENCRYPTION_KEY_ALGO_SETTING.get(settings);
+
+        try {
+            encryptionKey = encryptionKey(systemKey, keyLength, keyAlgorithm);
+        } catch (NoSuchAlgorithmException nsae) {
+            throw new ElasticsearchException(COULD_NOT_LOAD_KEY_MESSAGE, nsae);
+        }
+
+        assertEncryptionKeyNotNull();
+    }
+
+    private static int getKeyLength(Settings settings) {
+        final int keyLength = ENCRYPTION_KEY_LENGTH_SETTING.get(settings);
+
+        if (keyLength % 8 != 0) {
+            throw new IllegalArgumentException("invalid key length [" + keyLength + "]. value must be a multiple of 8");
         }
+
+        return keyLength;
+    }
+
+    private void assertEncryptionKeyNotNull() {
         assert encryptionKey != null : "the encryption key should never be null";
     }
 
+    private static SecretKey createSecretKey(char[] systemKey) {
+        final int keySizeBytes = KEY_SIZE / 8;
+        final ByteBuffer byteBuffer = StandardCharsets.ISO_8859_1.encode(CharBuffer.wrap(systemKey));
```

Comment:
Seems odd to be using IS_8859_1 for this. Should we be using base64 encoding/decoding instead? Otherwise we might wind up with control characters that are difficult/impossible to put into a settings string, right?

Reply:
Keith and I chatted offline. I'll change this to use base64 decoding (that's already how cloud will do it).

I think we'll still want to do
```
final ByteBuffer buffer = StandardCharsets.UTF_8.encode(CharBuffer.wrap(systemKey));
byte[] rawKey = Base64.getDecoder().decode(byteBuffer);
```

That way we can avoid creating a new String.

I'll make that change.